bgpd/bgp_packet.c

```diff
@@ -2886,9 +3024,20 @@ static int bgp_capability_msg_parse(struct peer *peer, uint8_t *pnt,
 					return BGP_Stop;
 			}
 			break;
+		case CAPABILITY_CODE_RESTART:
+			if ((hdr->length - 2) % 4) {
+				zlog_err("%pBP: Received invalid Graceful-Restart capability length %d",
+					 peer, hdr->length);
+				bgp_notify_send(peer, BGP_NOTIFY_CEASE,
+						BGP_NOTIFY_SUBCODE_UNSPECIFIC);
+				return BGP_Stop;
+			}
+
+			bgp_dynamic_capability_graceful_restart(pnt, action,
```

Comment:
Don't we need to pass in length?  What happens if the length specified is a multiple of 4 but the buffer size is not congruent with the length?

Reply:
AFAIU, the length is already checked above in `bgp_capability_msg_parse()`:
```c
hdr = (struct capability_header *)(pnt + 1);
```
We pass `hdr` to `bgp_dynamic_capability_graceful_restart()`, which checks against `hrd->length` (the actual size of the capability).